memory/allocator.go

```diff
@@ -235,6 +247,76 @@ func (a *ResourceAllocator) allocator() memory.Allocator {
 	return a.Allocator
 }
 
+type GcAllocator struct {
+	*ResourceAllocator
+}
+
+func (a *GcAllocator) Allocate(size int) []byte {
+	bs := a.ResourceAllocator.Allocate(size)
+	if len(bs) > 0 {
+		l := len(bs)
+		runtime.SetFinalizer(&bs[0], func(b *byte) {
+			if l != 0 {
+				// Allows us to reconstruct the slice without creating a circular dependency between
+				// the finalizer and the slice
+				bs2 := unsafe.Slice(b, l)
+				a.ResourceAllocator.Free(bs2)
+				l = 0
+			}
+		})
+	}
+	return bs
+}
+
+func (a *GcAllocator) Reallocate(size int, b []byte) []byte {
+	bs := a.ResourceAllocator.Reallocate(size, b)
+
+	// If the reallocation extended `b` the previous finalizer are still attached
+	if len(bs) > 0 {
+		l := len(bs)
+		runtime.SetFinalizer(&b[0], nil)
+		runtime.SetFinalizer(&bs[0], func(b *byte) {
+			if l != 0 {
+				// Allows us to reconstruct the slice without creating a circular dependency between
+				// the finalizer and the slice
+				bs2 := unsafe.Slice(b, l)
+				a.ResourceAllocator.Free(bs2)
+				l = 0
+			}
+		})
+	}
+
+	return bs
+}
+
+func (a *GcAllocator) Free(b []byte) {
+
+}
+
+func (a *GcAllocator) Account(size int) error { return a.ResourceAllocator.Account(size) }
+
+func (a *GcAllocator) MaxAllocated() int64 {
+	return a.ResourceAllocator.MaxAllocated()
+}
+
+func (a *GcAllocator) TotalAllocated() int64 {
+	return a.ResourceAllocator.TotalAllocated()
+}
+
+// Forces "all" unreachable memory to be garbage collected
```

Comment:
The quotes around _all_ are doing some work here, per the interior comments :laughing: